### DISCUSSION TITLE
Why did the ping/pong mechanism in the socket.io protocol get reversed?

### DISCUSSION THREAD

**Discussion by lazercorn** [socketio/socket.io#3983] [Q&A]
So the current protocol the ping/pong mechanism is reversed unlike older versions. The server issues a ping and the client responds with a pong. Isn't it better for the client to issue a ping and server respond with pong for scalability? Why was it changed?

and another question, since which version was this change implemented?

**Comment by lazercorn** [comment]
Ok. I just googled and apparently the reason is "a lot of reports from users that experience random disconnects due to ping timeout, After analysis, it seems to be caused by delayed timers on the client-side." according to this link: https://socket.io/blog/engine-io-4-release/

But I don't understand why switching to have server send ping would solve this. If this delayed timer problem is actually the case, wouldn't it be an issue in the javascript language itself or the browser, not with the socket.io protocol?

**Reply by lazercorn** [reply]
@darrachequesne

**Comment by darrachequesne** [comment]
Hi!

> Isn't it better for the client to issue a ping and server respond with pong for scalability?

I don't think it really matters in terms of performance. The only difference is an additional timer on the server side (one for the ping interval, and one for the ping timeout).

> Why was it changed?

Yes, we had received reports of random ping timeouts due to delayed timers. This problem was exacerbated by a low `pingTimeout` default value in `socket.io@3.0.0` (5s, which was increased to 20s in v4).

Related:

- https://github.com/socketio/socket.io/issues/3259
- https://github.com/socketio/socket.io/issues/3191
- https://github.com/socketio/socket.io/issues/2924
- https://github.com/socketio/engine.io/issues/312

> since which version was this change implemented?

This change was implemented in https://github.com/socketio/engine.io/commit/31ff87593f231b86dc47ec5761936439ebd53c20, included in `engine.io@4.0.0` and `socket.io@3.0.0`.

Reference: https://socket.io/docs/v4/migrating-from-2-x-to-3-0/

**Reply by lazercorn** [reply]
With regards to the server sending out pings, is each ping sent out in its own greenthread? Because if not, I can imagine if the server had to send out say 10k pings in one go, that would essentially block the OS thread and cause notable delay.

**Reply by darrachequesne** [reply]
The pings are not sent in bulk, each one has its own timer.

The code can be found here: https://github.com/socketio/engine.io/blob/7e9e544a9ac4365ef055c9482cc1e9fed624031c/lib/socket.js#L146-L156
